Wip v2.01

maxwellbenton opened the pull request.

Commit:
Rework readme

Commit:
Edit readme content

Commit:
Clarified and simplified README

Commit:
Remove bizaare Hash structure of cart

Commit:
Add specs for find_item_by_name_in_collection

This is a helper method that helps with the
location of a Hash within an Array. Students don't
yet know Enumerables or Hash-pair iteration, so
they really need this "boost."

Commit:
Remove Enumerable use in specs

Inappropriate for our laziness and (to be fair)
DRY-ness to needlessly confuse the matter.

Commit:
Consequence of de-Hashing the carts

Commit:
De-DRY before-blocks

Confusing for learners at this stage. But the
WETness does make me sad.

Commit:
Formatting

Commit:
Remove purchase-penalizing test

Seems real strange. If someone buys > coupon count
of items, cool, we should just charge them coupon
price and regular price. That seems the whole
point of keeping track of count in the
consolidated cart. If you don't meet the number,
yes, you should not get the coupon's power. But
more...seems wrong.

Succinctly: What grocer is going to complain that
people bough _more_?

Commit:
Remove inappropriate spec fanciness

Commit:
Remove test where multiple coupons come round

We've exercised them enough! :)

It's also hard to do without Enumerables so let's
cut that complexity.

Commit:
Clean up use of apply_clearance

It was unclear whether apply_clearance was
mutating or not. In the README we now say,
"mutate," so adjust specs to not require a return
value for success.

Also remove Enumerable fanciness.

Also, use a better matcher for testing Floats.

Commit:
Remove fanciness

Commit:
Remove to receive calls

This is so flakey and confusing, I think we're
better off, for a beginner audience not working in
a `class` context simply documenting the
expectation.

We set the expectation in both comments and the
README.

Commit:
Remove Hash-based thinking

Commit:
Remove stray receiver test

Commit:
Remove Hash-structure based calls

Commit:
Remove illegit. helper find_coupon

Commit:
Clarify retval non-important

Commit:
rm unused vars

Commit:
Weird case test

Commit:
Formatting

Commit:
Remove illegit find_item helper

Commit:
rm confusing Array with defalu value invocations

Commit:
Baseline grocer impl file

Commit:
Clarify learning goals

Commit:
Grammar

maxwellbenton committed:
fixes some grammar, clarifies checkout instruction

maxwellbenton committed:
fix code snippet errors

maxwellbenton commented on spec/grocer_spec.rb:
```diff
@@ -22,296 +23,311 @@
-        coupons_applied = apply_coupons(consolidated, [])
-        clearance_applied = apply_clearance(coupons_applied)
+        apply_coupons(consolidated, [])
+        apply_clearance(consolidated)
```
these lines and 257-259 were causing some problems while testing this - I've changed them to match the expected (previously, these methods returned values that were then passed on to the next, but now the lab requires students to modify the existing array, which would cause the other `apply_clearance` tests to pass but fail when using the method here or vice versa.

It looks though like these aren't being used, or am I wrong? The expectation is just running `checkout`, and not using whatever these lines are calculating to compare

sgharms replied:
I don't think that's right. In both of these items, the method is updating what's passed in.

https://github.com/learn-co-curriculum/ruby-enumerables-hash-practice-green-grocer-lab/blob/solution/grocer.rb#L17-L36

https://github.com/learn-co-curriculum/ruby-enumerables-hash-practice-green-grocer-lab/blob/solution/grocer.rb#L38-L45

maxwellbenton replied:
the instructions in the new starter code are to modify the existing array and the 2.01 solution for `checkout` modifies the cart without returning anything: https://github.com/learn-co-curriculum/programming-univbasics-nds-green-grocer/blob/wip-v2.01-solution/grocer.rb

It is possible with this set up to pass the `apply_clearance` tests without having to return anything, but then fail the tests here because they're expecting a return

maxwellbenton replied:
so this aligns with the new solution

maxwellbenton replied:
Also other parts of the tests: https://github.com/learn-co-curriculum/programming-univbasics-nds-green-grocer/blob/wip-v2.01/spec/grocer_spec.rb#L76-L87

maxwellbenton commented on README.md:
```diff
@@ -54,111 +142,140 @@ then the method should return the hash below:
+  * `Array`: a collection of item `Hash`es
+  * `Array`: a collection of coupon `Hash`es
+* Returns:
+  * The first `Array`, ***updated***. We ***should*** be editing the item
```
I'm a little concerned requiring the modification of the existing array has made this more difficult. Students are very primed for methods that take in something, rearrange it into a new thing and return that thing and may need additional cues in the readme here to guide them, especially since consolidate cart _does_ return a new array

sgharms replied:
In the previous version, we implied that we used return values from methods that did in-place updates.

e.g. `apply_coupons()`

`@avocado_result = apply_coupons(@consolidated_cart, [@avocado_coupon])`

https://github.com/learn-co-curriculum/ruby-enumerables-hash-practice-green-grocer-lab/blob/master/spec/grocer_spec.rb#L57

But...

https://github.com/learn-co-curriculum/ruby-enumerables-hash-practice-green-grocer-lab/blob/solution/grocer.rb#L17

maxwellbenton replied:
Ah I see. I think it can stay as is, but just needs some sort of explanation calling out the difference - maybe an example refresher in the readme

maxwellbenton reviewed (commented):
Pushed up some fixes to the readme and a few for the tests. I think overall, it is better, butI think asking students to pass an array into a method, modify it without returning anything and then use that array later might be pretty novel to students and could make this more difficult than before.

I think it can be kept, but may need some more grease in the readme.

sgharms commented on spec/grocer_spec.rb:
```diff
@@ -95,7 +96,7 @@
         apply_coupons(consolidated_cart, [avocado_coupon])
         found_item = find_item_by_name_in_collection(item_with_coupon_applied_name, consolidated_cart)
-        expect(found_item[:price]).to eq(2.50), "After applying a $5 for 2 coupon to avocadoes, the price per unit is 2.50"
+        expect(found_item[:price]).to eq(2.50)
```
Why cut the error text?

maxwellbenton replied:
if there is no text, the error produces a
```
expected: 2.50
actual: nil # or whatever the result of their work is
``` 

With the text, they get that message and can't see the actual result of their work. Could fixed with interpolation.

sgharms commented on spec/grocer_spec.rb:
```diff
@@ -251,11 +252,11 @@
         beer = find_item_by_name_in_collection('BEER', items)
         cart = [find_item_by_name_in_collection('BEETS', items), beer, beer, beer]
-        c = [coupons[1]]
+        coupon = [coupons[1]]
```
Not accurate. This is an array. Calling it coupons clashes with the let() statement up top (bitter, bitter experience)

maxwellbenton replied:
oh my mistake. It can go back, I just wanted something there that was spelled out

Commit:
Add expec: find_item_by_name_in_collection

Test the positive case

Commit:
Wip no inplace methods (#2)

* Standardize Spec: apply_coupon()'s variable

Make things easier by having apply_coupon() always
return coupon_applied_cart().

* Remove unused call to apply_coupons

* Standardize spec: consolidated_cart

Make tests easier to read for beginners. Name
consolidate_cart()'s retvalue consistently.

* Standardize calls to apply_clearance

* Simplify checkout methods

* w/ coupon -> W/COUPON to match tests
